independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/QuarkusClassLoader.java

```diff
@@ -72,9 +72,16 @@ public static boolean isClassPresentAtRuntime(String className) {
      * Indicates if a given class is considered an application class.
      */
     public static boolean isApplicationClass(String className) {
+        final ClassLoader ccl = Thread.currentThread().getContextClassLoader();
+        if (!(ccl instanceof QuarkusClassLoader)) {
+            throw new IllegalStateException("The current classloader is not an instance of "
+                    + QuarkusClassLoader.class.getName() + " but " + ccl.getClass().getName());
+        }
+
         String resourceName = fromClassNameToResourceName(className);
-        List<ClassPathElement> res = getElements(resourceName, true);
-        return !res.isEmpty();
+        ClassPathResourceIndex classPathResourceIndex = ((QuarkusClassLoader) ccl).getClassPathResourceIndex();
+
+        return classPathResourceIndex.getFirstClassPathElement(resourceName) != null;
```

Comment:
This can be simplified a tad by using pattern matching with `instanceof`

Reply:
Why not. I don't mind adding some cuteness but I would like to get this patch in and avoid iterating on cosmetic details. I don't think the code is less readable than it used to be.

Reply:
Sure, I'm just mentioning it because the whole point of the commit was to simplify the method